#2646 StringShrinkerWithMin generates samples based on source-Arb

aSemy opened the pull request:
fixes #2646

- `StringShrinkerWithMin` will select a random 'simplest' char, instead of always using `a`, based on the provided codepoints.
- Added KDoc to highlight `StringShrinkerWithMin.simplestChar` 
- `respect min value` test needed updating
- Added a couple of tests to check that samples only contain codepoints of the source Arb

aSemy committed:
#2646 StringShrinkerWithMin generates samples based on source-Arb codepoints

aSemy commented on kotest-property/src/commonMain/kotlin/io/kotest/property/arbitrary/strings.kt:
```diff
@@ -86,11 +87,30 @@ object StringShrinker : Shrinker<String> {
+ */
 class StringShrinkerWithMin(
    private val minLength: Int = 0,
    private val simplestChar: Char = 'a',
```
I think to be more clear and avoid accidental usage, the primary constructor with the `simplestChar` should be converted to a secondary constructor, and then `@Deprecated`.  The primary, and only, constructor should accept `codepoints: Arb<Codepoint>`, as it is more useful than a static character.

Thoughts?

sksamuel replied:
Do we want shrinks to be random? wouldn't it be better if a shrunk value is always deterministic ?

aSemy replied:
Umm good question. I'm not sure, I hadn't thought about it and I don't have as good overview to answer it to be honest. My instinct would be that it's fine, and allowing for Arbs in shrinkers to make them random is okay. But maybe there's benefit to them being completely deterministic. Or there should be a principle that shrinkers should be deterministic.

If this PR doesn't fit then that's okay :)

sksamuel replied:
I like the idea but I think shrinks should be deterministic.
Like to be, 'a' is the best candidate for a-z shrinking. Perhaps there's a way to always choose the "best" shrink given a list of chars?

aSemy replied:
How about this: The simplest char could be the first character of the pre-shrink String?

```kotlin
   override fun shrink(value: String): List<String> {

      val isShortest = value.length == minLength

      val simplestChar: Char? = value.firstOrNull()
      val isSimplest = simplestChar == null || value.all { it == simplestChar }
      // ...
```

aSemy replied:
I've updated the PR to reflect my suggestion. I hope it's something that works! 

The old behaviour of a static 'simplest char' can still be retained, if necessary. I've written, but commented-out, an example secondary constructor with a `@Deprecated`.

It's not ready for merging as I haven't written tests, just updated the existing ones. In particular there's no testing for when `simplestCharSelector` returns `null`.

sksamuel replied:
Sorry for delay in replying.

I think that the simplest char being the first char in the input string could work.
It would be repeatable for any given input at least.

Another option is to introduce some kind of Codepoints/Alphabet interface which specifies the low point for a given range of chars ?

aSemy replied:
No problem! I hope you had a good time over Christmas and New Years.

I'm not quite sure how an `Alphabet` interface would work, it would help to see an example. 

How should the 'simplest' char be selected? Would users know what this would mean? Would the 'simplest char' be relevant to anything else, or only `StringShrinker`? If 'simplest char' is only relevant to `StringShrinker` then it should only be defined and used there, so the context is clearer.

What happens if an `Alphabet` needs to be filtered or mapped - would the simplest char need to be defined on each operation? 

When defining a 'simplest char', does it have to be contained within the `Alphabet`? What happens if it isn't?

aSemy replied:
@sksamuel  I've added the tests for `simpleCharSelector`. Aside from the deprecation annotations and backwards compatibility, I think it's good to go! Let me know if there's anything more I can do.

aSemy committed:
move stringArb out of the arbSample constructor, clarify comments, add clue

aSemy committed:
oops, fix assertion

aSemy committed:
Merge branch 'master' into feature/2646-string-shrink-limit-simplest-char

aSemy committed:
update StringShrinkerWithMin to select a simplestChar from a pre-shrunk String, so it will be valid and deterministic

aSemy committed:
return the variant-selection to match the original

aSemy committed:
Merge branch 'master' into feature/2646-string-shrink-limit-simplest-char

aSemy committed:
add javadoc, undo formatting

aSemy committed:
actually, it's simpler with sequential 'if's, as it covered all combinations of is/isn't simplest/shortest better than the 'when' statement

aSemy committed:
Merge remote-tracking branch 'origin/master' into feature/2646-string-shrink-limit-simplest-char

aSemy committed:
Merge remote-tracking branch 'origin/master' into feature/2646-string-shrink-limit-simplest-char

aSemy committed:
test StringShrinker's simplestCharSelector

(additionally, make the test names consistent and slightly rearrange StringShrinkerWithMin statements for clarity)

sksamuel commented on kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/StringShrinkerWithMinTest.kt:
```diff
@@ -95,44 +106,184 @@ class StringShrinkerWithMinTest : DescribeSpec({
+Shrink #6: "ffffc ffff" fail
+Shrink #7: "fffff ffff" fail
+Shrink #8: "ffffffffff" fail
+Shrink result (after 8 shrinks) => "ffffffffff"
```
Why did it end up with f's when a was the first char in the arg ?

sksamuel replied:
never mind I see this test is for a custom char selector.
